### PR TITLE
Privacy Settings

### PR DESCRIPTION
**Pull request by judtinzhang**
This PR is to introduce Privacy into Penn Labs products. The original intention was for Mobile products wherein we are able to access student demographic information and academic identity, however this PR can easily be extended to all products with privacy concerns. To move forward with Mobile plans on incorporating user data into our product, we want to ensure that users have the ability to opt-out on command.  

- Create `PrivacySetting` model to store privacy settings for a particular user.
- Create `PrivacyResource` model to store all resources that Penn Labs allows users reserve the right to withhold from Labs products.
- Create a view for list and update routes to access and change privacy settings

TODOS:
- [x] Create the `PrivacySetting` model
- [x] Create a post-save hook to create privacy settings for new users
- [x] Create a post-save hook for new privacy resources
- [x] Create PATCH/GET view
- [x] Test cases

### CODE REVIEW

File: backend/accounts/models.py [hunk 1]
```diff
@@ -133,3 +133,53 @@ class PhoneNumber(models.Model):
...
+@receiver(post_save, sender=User)
+def load_privacy_settings(sender, instance, created, **kwargs):
+    """
+    This post_save hook triggers automatically when a User object is saved, and loads in default
+    privacy settings for the User
+    """
+
+    # In most cases, first checking if settings exists should reduce the number of queries
+    # to the database
+    if not PrivacySetting.objects.filter(user=instance).exists():
```
Comment (tuneerroy) [inline backend/accounts/models.py:181]: would checking if `instance.privacy_setting` exists work instead, here?

**Review by tuneerroy — APPROVED** [review]

File: backend/accounts/models.py (hunk 1 above)
Comment (judtinzhang) [reply backend/accounts/models.py]: Agreed!

File: backend/accounts/models.py [hunk 2]
```diff
@@ -133,3 +133,53 @@ class PhoneNumber(models.Model):
...
+def add_privacy_resource(sender, instance, created, **kwargs):
+    """
+    This post_save hook triggers whenever a new privacy resource is added.
+    Each User will receive a new privacy setting with this resource, enabled
+    to true.
+    """
+    users = User.objects.all()
+    resources = [None] * users.count()
+    for i, user in enumerate(users):
+        resources[i] = PrivacySetting(user=user, resource=instance, enabled=True)
```
Comment (rohangpta) [inline backend/accounts/models.py:157]: Would this also work? 

```py
resources = [PrivacySetting(user=user, resource=instance, enabled=True) for user in users]
```

File: backend/accounts/models.py [hunk 3]
```diff
@@ -133,3 +133,53 @@ class PhoneNumber(models.Model):
...
+    """
+    This post_save hook triggers automatically when a User object is saved, and loads in default
+    privacy settings for the User
+    """
+
+    # In most cases, first checking if settings exists should reduce the number of queries
+    # to the database
+    if not instance.privacy_setting.exists():
+        for resource in PrivacyResource.objects.all():
+            PrivacySetting.objects.create(
```
Comment (rohangpta) [inline backend/accounts/models.py:183]: Can you refactor to use `bulk_create`?

File: backend/accounts/views.py [hunk 4]
```diff
@@ -527,3 +528,19 @@ def post(self, request, format=None):
...
+
+
+class PrivacySettingView(
+    generics.GenericAPIView, mixins.ListModelMixin, mixins.UpdateModelMixin
+):
+    serializer_class = PrivacySettingSerializer
+    permission_classes = [IsAuthenticated]
+
+    def get_queryset(self):
+        return self.request.user.privacy_setting.all()
```
Comment (rohangpta) [inline backend/accounts/views.py:540]: You might want to use `select_related` and fetch privacy resources here, to avoid an N+1 while serializing.

**Comment by rohangpta** [comment]
This looks great! Thank you for doing this -- I left a couple comments.

I'm curious, what kind of data are we thinking of using? Are users opted-in by default? Do we have to declare our privacy policy to Apple?

**Comment by rohangpta** [comment]
@Clue88 would be great if you could read this, lots of things that you can pick up from this PR :) Feel free to drop questions in this chain if you have any

File: backend/accounts/models.py (hunk 2 above)
Comment (judtinzhang) [reply backend/accounts/models.py:157]: Oh wow I just looked it up and yeah I thought pre-allocating the size of the list would be more efficient, but turns out list comprehension does in under the hood as well. Thanks!

File: backend/accounts/views.py (hunk 4 above)
Comment (judtinzhang) [reply backend/accounts/views.py:540]: Thanks for the catch!

**Comment by judtinzhang** [comment]
@rohangpta We were thinking of things such as scraping Path@Penn info (ex. allowing Penn Labs to access user course info, academic info, housing, etc.) and also some Portal features (allowing us to anonymize their voting data and send it back to the poll creator).

**Pull request merged**